Add Solar-Log platform

Ernst79 opened the pull request:
## Description:
Adding Solar-Log platform for Solar-Log (PV monitoring) devices. 

## Example entry for `configuration.yaml` (if applicable):

`sensor:`
`  platform: solarlog`

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

Ernst79 committed:
Add Solar-Log sensor

homeassistant commented:
Hi @Ernst79, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#27036).

Once you do that we will be able to review and accept this pull request.

Thanks!

Ernst79 committed:
Codeowners update

balloob commented on script/gen_requirements_all.py:
```diff
@@ -165,6 +165,7 @@
     "sleepyq",
     "smhi-pkg",
     "solaredge",
+    "sunwatcher",
```
Don't add it here, this is only for integrations that have tests.

Ernst79 replied:
Removed

balloob commented:
Hey Ernst, thanks for your first contribution. Welcome to the Home Assistant community! 🐬

balloob commented on homeassistant/components/solarlog/manifest.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "domain": "solarlog",
+  "name": "Solar-Log",
+  "documentation": "https://www.home-assistant.io/components/solarlog",
```
We just changed all the urls ! :) 

```suggestion
  "documentation": "https://www.home-assistant.io/integration/solarlog",
```

Ernst79 replied:
Changed

balloob commented on homeassistant/components/solarlog/sensor.py:
```diff
@@ -0,0 +1,233 @@
+    {
+        vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): cv.time_period,
```
Let's not make the scan interval configurable, instead let's just set it to a reasonable default.

Ernst79 replied:
fixed. I've removed the config option and made a fixed scan interval of 1 minute.

balloob commented:
May I suggest that you use a config flow instead of platform configuration? You can do it as follows:

You can do it easily with our new scaffolding tool. From your dev environment, run:

```bash
python3 -m script.scaffold config_flow
```

It will generate a config flow and tests that pretty much match your schema 👍 Then change `setup_platform` into `async_setup_entry`.

Ernst79 committed:
Update homeassistant/components/solarlog/manifest.json

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

Ernst79 committed:
remove sunwatcher from gen_requirements_all.py

Ernst79 committed:
remove sunwatcher from requirements_test_all.txt

Ernst79 committed:
Remove scan_interval as configuration variable

I've set it to a fixed scan_interval of 1 minute. Removed the configuration option.

Ernst79 committed:
Fix black format

Ernst79 commented:
> May I suggest that you use a config flow instead of platform configuration? You can do it as follows:
> 
> You can do it easily with our new scaffolding tool. From your dev environment, run:
> 
> ```shell
> python3 -m script.scaffold config_flow
> ```
> 
> It will generate a config flow and tests that pretty much match your schema 👍 Then change `setup_platform` into `async_setup_entry`.

Good suggestion. I do like to make this change in the future, but as I'm not a programmer, so I need time to look into this and understand what I'm doing. I would propose that we first implement the current version of the integration and that I will add this in a future update.  

I also have issues with the 'git fetch upstream dev' command, so I need to setup my dev environment again. It's complaining about 'unable to find remote helper for https', and all solutions on google don't seem to help me. So that's makes it impossible to update things from my local git.

Ernst79 committed:
Config flow added (__init__.py)

Ernst79 committed:
Config flow added (manifest.json)

Ernst79 committed:
Config flow added (const.py)

Ernst79 committed:
Config flow added (config_flow.py)

Ernst79 committed:
Config flow added (strings.json)

Ernst79 committed:
Config flow added (en.json translation)

Ernst79 committed:
Config flow added (sensor.py rewritten)

Ernst79 committed:
Config flow added (sensor.py)

Ernst79 committed:
Config flow added (config_flows.py)

Ernst79 committed:
resolve conflict config_flows.py

Ernst79 commented:
@Balloob I've updated the code to have both the config flow, as you suggested, and the possibility to use the old fashioned platform configuration (yaml). It was a little bit more work than expected, but it's working now. Github is saying that there is a conflict in the config_flows.py file, but it cannot be resolved by me, I guess. Not sure what I'm doing wrong.

balloob committed:
Merge branch 'dev' into solarlog

balloob commented:
Resolving the conflict was easy, I just pressed the "resolve conflict" button for you 👍 

Did you run `python3 -m script.scaffold config_flow` ? It looks like you have diverted quite a bit from the original template, including the tests are missing.

Ernst79 committed:
Add tests

Ernst79 committed:
add tests

Ernst79 committed:
add tests

Ernst79 commented:
@balloob I've added the tests. Yes, I've differted a bit from the template, as I like to be able to use the yaml config as well. I've used the cert_expire sensor as a template and moved a part of sensor.py to const.py. 

I don't know how to fix the failing codecov test?
And the documentation is not really clear if I need to include solarlog/sensor.py, solarlog/* or nothing in .coverage. What Shoud I use?

Ernst79 committed:
Update .coverage to include all files for solarlog

balloob commented on homeassistant/components/solarlog/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    hass.async_create_task(
```
Please await the unload.

```python
return await hass.config_entries.async_forward_entry_unload(…)
```

Ernst79 replied:
Fixed.

balloob commented on homeassistant/components/solarlog/config_flow.py:
```diff
@@ -0,0 +1,85 @@
+        """Check if we can connect to the Solar-Log device."""
+        host = user_input.get(CONF_HOST, DEFAULT_HOST)
+        try:
+            SolarLog(f"http://{host}")
```
This is doing I/O and so cannot be run inside an async function. Run it in the executor instead:

```
await self.hass.async_add_executor_job(SolarLog, f"http://{host}")
```

Ernst79 replied:
Fixed

balloob commented on homeassistant/components/solarlog/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "user": {
+        "title": "Define your Solar-Log connection",
+        "data": {
+          "host": "The hostname or ip-address of your Solar-Log device (without http://)",
```
Can we see if `http://` is added inside the config flow, and add it if it is not ?

Ernst79 replied:
Good idea. I’ve added this.

balloob commented on homeassistant/components/solarlog/strings.json:
```diff
@@ -0,0 +1,21 @@
+        "title": "Define your Solar-Log connection",
+        "data": {
+          "host": "The hostname or ip-address of your Solar-Log device (without http://)",
+          "name": "The prefix to be used for your Solar-Log sensors"
```
Can we query this from the Solar-Log installation and thus omit from the form?

Ernst79 replied:
Just checking if I understand you right. Do you mean that you want to remove this from the (config_flow) form, so only ask for the host? Is it ok to keep the configuration option for ‘name’ in the configuration.yaml?

balloob commented on homeassistant/components/solarlog/sensor.py:
```diff
@@ -0,0 +1,150 @@
+        return
+
+    # Create solarlog data service which will retrieve and update the data.
+    data = SolarlogData(hass, api, host)
```
If this is doing I/O, run it in executor.

Ernst79 replied:
Fixed

balloob commented on homeassistant/components/solarlog/sensor.py:
```diff
@@ -0,0 +1,150 @@
+    host = entry.data[CONF_HOST]
+    platform_name = entry.title
+    try:
+        api = SolarLog(f"http://{host}")
```
This is doing I/O, run it in executor.

Ernst79 replied:
Fixed

Ernst79 committed:
Fix await the unload

Ernst79 committed:
Adjust icons, add http:// to default host

Ernst79 committed:
Change icons

Ernst79 committed:
Add http:// to host if not provided, fix await

Ernst79 committed:
Add http:// to host if not provided, fix await

Ernst79 committed:
Adjust tests for http:// added to host

Ernst79 committed:
remove line

Ernst79 committed:
Remove without http:// requirement

Ernst79 committed:
Remove without http;// requirement

Ernst79 commented:
@balloob All remarks should be resolved now. Also updated the documentation. Only the remark about query-ing the name from the Solar-Log connection is not possible, I think, see my comment on that question above.

balloob reviewed (approved):
Awesome! 🎉  Great first contribution 🐬  Welcome to the Home Assistant community 👋
